Bkplot

cdf6gc opened the pull request:
Uploaded a method to create a basic plot following Burtin's antibiotic template. The python file has a command at the end to call the method for the CO analysis file.

The file plots all parameters who S1 and ST values exceed the minimum values fed by the user. I am still working on the axis scaling but this should serve as a basic method to build off of and add widgets to.

cdf6gc committed:
Adding Script for Basic Burtins Plot

cdf6gc committed:
updated filepathway

houghb commented:
I am about to start looking at this code, but my first comment is that all of our main modules (plotting, analysis, etc) should be in /HDSAviz/HDSAviz/.  Right now this is in /HDSAviz/PlottingScripts/, and from this directory it cannot be called from other parts of the program.  If we have enough plotting scripts that they need their own folder (unlikely), then we can make another folder at the same level as the existing `tests` folder.

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+from bokeh.plotting import figure, show, output_file
+
+
+"""This script is just the first step in visualizing the data set. This file
```
Move this to the main comment block at line 1

cdf6gc replied:
Will do

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+"""
+
+
+def makeplot(Dataframe, Cols, minvalues):
```
We haven't talked about this as a group yet, but I think we should settle on a consistent format for naming and documenting functions.  I have been using `verb_noun()` for my convention, so I would name this `make_plot()`.  It doesn't matter what format we choose, let's just talk about it and choose one.

Also, we should be consistent in how we format the docstrings of functions.  I start out with a short description, then use:

```
Parameters:  
---------------
param1 : description of it including the object type
param2 : 

Returns:  
-----------
thing : description of what is returned
```

Again, it doesn't matter what we do, let's just talk about this and be consistent.

cdf6gc replied:
I agree.

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+
+    This is the method to generate a bokeh plot followung the burtin example
+    template at the bokeh website. For clarification, parameters refer to an
+    output being measured (Tmax, C, rxn2, etc.) and stats refer to the 1st or
```
In the functions that I've written to read and process all our input data I've updated the parameters to say "k2" instead of "rxn2" to be more specific about what was adjusted.  We should update this throughout our code and documentation.

cdf6gc replied:
Changed

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+import pdb
+
+from bokeh.plotting import figure, show, output_file
+
```
Please add `import data_processing as dp` so we can use the new functions I wrote for getting the pandas dataframes

cdf6gc replied:
I am adding the changes that rely on your files as comments because I need to merge with your branch first before it will take effect. But it should hopefully be a quick change when we merge branches

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+    Generates an html of the plot. The axis is a log scale.
+    """
+    # Read in csv file as panda dataframe.
+    tdf = pd.read_csv(Dataframe, delimiter=' ', skipinitialspace=True,
```
The Dataframe parameter of your function should be updated to simply **be** the dataframe that you define as tdf.  I have a function (`get_sa()`) that will read all the analysis files and clean/build the pandas dataframes for all output measures.  That function will provide the cleaned dataframe to use as input for your makeplot.  As far as specific changes to your code here this would mean deleting line 35 and just changing line 36 to `df = Dataframe` (and updating the function documentation above accordingly).

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+implement the method
+"""
+
+
```
You can add default values to cols `=['S1', 'ST']` because that will almost always be how we plot it and it will make plotting function calls shorter.

cdf6gc replied:
Set as default values

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+    if maxval < 0.1:
+        labels = np.delete(labels, 0)
+    # Size of radii of each label
+    pdb.set_trace()
```
Why is this here?

cdf6gc replied:
Removed

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+    p.text([-15, -15, -15], [18, 0, -18], text=Cols,
+           text_font_size="9pt", text_align="left", text_baseline="middle")
+
+    output_file("Sensitivity.html", title=Dataframe)
```
Two points on the output file:
-  I think we should save plots outside the repository (perhaps next to the data directory?)
-  Can you make another parameter for the function that indicates whether or not a plot should be saved?  Often a user will want to explore a bunch of plots in an interactive session, but doesn't want to save them all until they get exactly the plot they would like.

cdf6gc replied:
Booked save tool is now enabled to take care of this

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+
+    # Determine wedge size based off number of parameters
+    big_angle = 2.0 * np.pi / (len(df)+1)
+    # Determine division of wedges for plotting bars based on # stats plotted
```
It is difficult to make this a stacked bar (instead of bars next to each other for S1 and ST)?  S1 will always be less than ST, so a stacked bar would save space and may make it easier to see the differences between S1 and ST...

cdf6gc replied:
I don't think making a stacked bar would be hard. The problem would be that since the scale is log the sizes of the outer bar would be skewed smaller compared to the inner bar.

houghb replied:
Maybe I am incorrectly using the term "stacked bar"...  If what you have right now is this (picture these oriented vertically):
first order :           -------------------
second order :     xxxxxxxxxxxxxxxx

Then I am suggesting this:
-------------------xxx

In this case I don't think the log scale should be a problem...  We could also talk about whether it needs to be log scale, but that is probably best explained in person

cdf6gc replied:
I think I understand what you are saying now. This should be possible

swapilpaliwal replied:
Many parameters have ST lower then S1 especially in low sensitivity region.
I don't know specific reasons though.
Please see attached image.

Also, is there a way we can make sorted plots ? either according to S1 or
ST as that would be easier to slice and dice once we explore more
interactive options like top 5 or top 20.

On Tue, Feb 23, 2016 at 4:03 PM, houghb notifications@github.com wrote:

> In PlottingScripts/Basic_Plot.py
> https://github.com/houghb/HDSAviz/pull/2#discussion_r53873048:
> 
> > -    colors = ["#0d3362", "#c64737", "black"]
> > -    stat_color = OrderedDict()
> > -    for i in range(0, Cols.size):
> > -        stat_color[i] = colors[i]
> > -    # Reset index of dataframe.
> > -    df = df.reset_index(drop=True)
> >   +
> > -    # Sizing parameters
> > -    width = 800
> > -    height = 800
> > -    inner_radius = 90
> > -    outer_radius = 300 - 10
> >   +
> > -    # Determine wedge size based off number of parameters
> > -    big_angle = 2.0 \* np.pi / (len(df)+1)
> > -    # Determine division of wedges for plotting bars based on # stats plotted
> 
> It is difficult to make this a stacked bar (instead of bars next to each
> other for S1 and ST)? S1 will always be less than ST, so a stacked bar
> would save space and may make it easier to see the differences between S1
> and ST...
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/houghb/HDSAviz/pull/2/files#r53873048.

houghb replied:
@swapilpaliwal can you put the image in Slack?  Nothing showed up here.  It shouldn't be the case that ST values are lower than S1 so I need to look into this.

cdf6gc replied:
I have seen instances where S1 is greater than ST (rxn107 in analysis_CO2.txt is one example). Ordering the the values shouldn't be too difficult because it should just require re-ordering the panda dataframe by a user specified column. Perhaps this would best be implemented as a separate function though.

houghb replied:
Ah, I see.  Looking at that example in particular it is clear that this is
a problem when the value of a sensitivity index is smaller than it's
confidence range and the value is close to zero.  This shouldn't be a
problem with any stacked bar type plots because differences that small
won't show, and even if they do then the error bars on the plot will
indicate what is going on.

On Wed, Feb 24, 2016 at 9:53 AM, Chris Fu notifications@github.com wrote:

> In PlottingScripts/Basic_Plot.py
> https://github.com/houghb/HDSAviz/pull/2#discussion_r53977810:
> 
> > -    colors = ["#0d3362", "#c64737", "black"]
> > -    stat_color = OrderedDict()
> > -    for i in range(0, Cols.size):
> > -        stat_color[i] = colors[i]
> > -    # Reset index of dataframe.
> > -    df = df.reset_index(drop=True)
> >   +
> > -    # Sizing parameters
> > -    width = 800
> > -    height = 800
> > -    inner_radius = 90
> > -    outer_radius = 300 - 10
> >   +
> > -    # Determine wedge size based off number of parameters
> > -    big_angle = 2.0 \* np.pi / (len(df)+1)
> > -    # Determine division of wedges for plotting bars based on # stats plotted
> 
> I have seen instances where S1 is greater than ST (rxn107 in
> analysis_CO2.txt is one example). Ordering the the values shouldn't be too
> difficult because it should just require re-ordering the panda dataframe by
> a user specified column. Perhaps this would best be implemented as a
> separate function though.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/houghb/HDSAviz/pull/2/files#r53977810.

houghb commented on PlottingScripts/Basic_Plot.py:
```diff
@@ -0,0 +1,128 @@
+    show(p)
+
+# Command call to implement example plot.
+makeplot('../../HDSAviz_data/analysis_CO.txt', np.array(['S1', 'ST']), np.array([0.001, 0.001]))
```
if you add the line `import data_processing as dp` in your import section, then this command to make an example plot will become:

```
sa = dp.get_sa()
makeplot(sa['CO'][0], np.array(['S1', 'ST']), np.array([0.001, 0.001]))
```

swapilpaliwal replied:
Blake can you include this as comment in your code as usage??

sa['CO'][0] seems a bit wierd and confusing to use on it own.

can we instead put it as an intuitive named_variable ??

On Tue, Feb 23, 2016 at 4:09 PM, houghb notifications@github.com wrote:

> In PlottingScripts/Basic_Plot.py
> https://github.com/houghb/HDSAviz/pull/2#discussion_r53873727:
> 
> > -    label_angle = np.array(-big_angle/2+angles)
> > -    label_angle[label_angle < -np.pi/2] += np.pi
> > -    # Placing Labels and Legend
> > -    p.text(xr, yr, df.Parameter, angle=label_angle,
> > -           text_font_size="9pt", text_align="center", text_baseline="middle")
> >   +
> > -    p.rect([-40, -40, -40], [18, 0, -18], width=30, height=13,
> > -           color=list(stat_color.values()))
> > -    p.text([-15, -15, -15], [18, 0, -18], text=Cols,
> > -           text_font_size="9pt", text_align="left", text_baseline="middle")
> >   +
> > -    output_file("Sensitivity.html", title=Dataframe)
> > -    show(p)
> >   +
> >   +# Command call to implement example plot.
> >   +makeplot('../../HDSAviz_data/analysis_CO.txt', np.array(['S1', 'ST']), np.array([0.001, 0.001]))
> 
> if you add the line import data_processing as dp in your import section,
> then this command to make an example plot will become:
> 
> sa = dp.get_sa()
> makeplot(sa['CO'][0], np.array(['S1', 'ST']), np.array([0.001, 0.001]))
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/houghb/HDSAviz/pull/2/files#r53873727.

houghb replied:
This seems pretty intuitive to me once you understand what the `get_sa()` function returns.  I can add more documentation, but it is already described in that functions docstring.  This statement means "give me the dataframe of the 1st order sensitivity analysis for CO".  `sa` is a dictionary where the keys are the names of the different output measures (CO is one of those measures).  The dictionary value for each key is a list of two items - the zero index of the list [0] is the first and total order sensitivity index dataframe, and the second index of the list [1] is the second order index dataframe (if it exists).  All of this is documented in the function, but I can work on making it clearer.

cdf6gc committed:
Deleted folder to add files to different directory

cdf6gc committed:
Updated plot with formatting corrections, others pending until team meeting

cdf6gc committed:
Added list of comments (un)addressed and updated plotting script with formatting changes

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+               x_range=(-420, 420), y_range=(-420, 420),
+               min_border=0, outline_line_color="black",
+               background_fill_color="#f0e1d2", border_fill_color="#f0e1d2")
+
```
# p = figure(*_args, *_*kwargs) can include a tools command

This is a place where tools = [('save')] type option can be put up to save the figure
Please see [Bokeh save tool](http://bokeh.pydata.org/en/0.10.0/docs/user_guide/tools.html#id16) if this works fine.

As a backup i have included a save plot command in interactive_vis_tool.py but that doesn't give user an option to name files.

cdf6gc replied:
Added in the tool bar. Save and Resize appear to be working fine even if the radii are hardcoded in resizing adjusts them when I ran it in your notebook.

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+import pdb
+
+from bokeh.plotting import figure, show, output_file
+
```
# Important import statement :

Please include import statement for some tools here like below.

from bokeh.models.tools import (BoxZoomTool, ResetTool,
                                PreviewSaveTool,
                                ResizeTool, 
                                PanTool, PolySelectTool, 
                                WheelZoomTool)

please also see [here](http://bokeh.pydata.org/en/0.10.0/docs/user_guide/tools.html#hovertool) for instructions on implementing tools like the hover tool or pinch and zoom that we were planning.

cdf6gc replied:
Enabled these basic tools.

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+    p.text([-15, -15, -15], [18, 0, -18], text=Cols,
+           text_font_size="9pt", text_align="left", text_baseline="middle")
+
+    return p
```
# General comment:

Overall, I loved using a version of this code but i had a real hard time finding out what are actual variables are describing magnitude of sensitivity and the corresponding parameter. 
I wanted to call those variables in hover tool in format like [here](http://bokeh.pydata.org/en/0.10.0/docs/user_guide/tools.html#basic-tooltips)

hover.tooltips = [
    ("index", "$index"),
    ("variable", "syntax for values" )
    ("sensitivity", "@radius"),
    ("foo", "@foo"),
    ("bar", "@bar"),
]

does anyone have any idea on @ and $ conventions???
Can we implement something like above for hover tool? 

if we understand this nomenclature, then we can also build sliders and other user interfaces directly in bokeh and get great web based widgets instead of notebook ones which we get by using ipywidgets.

swapilpaliwal replied:
# Comment about returning values.

Whenever developing a plotting function please return values like here in last line without plots() or show() like commands  and call these plotting or visualization commands outside the function block so that the figure or plots features can be used in raw format.

bar replied:
@swapilpaliwal uhh?

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+    # Reset index of dataframe.
+    df = df.reset_index(drop=True)
+
+    # Sizing parameters
```
# It is going to be Difficult but for ideas :

### can sizing parameters be made automatic so as to adjust based on window size??

There is a [resizing tool in bokeh](http://bokeh.pydata.org/en/0.10.0/docs/user_guide/tools.html#resizetool) and javascript shall probably make it flexible to interact with resizing..... 

Default values are definitely helpful but might be  trouble

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+    p.text(0, radii[:], [str(r) for r in labels[:]],
+           text_font_size="8pt", text_align="center", text_baseline="middle")
+    # Plot the values of each stat for each parameter.
+    for c in range(0, Cols.size):
```
# Note :

can expressions like """ (((np.log10(df[Cols[c]] / + labels[0])) + labels.size) \* +  (outer_radius - inner_radius) /  +  labels.size + inner_radius) """ be given an understandable variable name ??? 
like may be radius_of_something .
It was hard to figure what's being plotted based on the expressions and c's  ......

houghb replied:
perhaps just a comment above these rad assignments is sufficient; if the variable names get too long that adds its own set of problems for readability.

swapilpaliwal commented on HDSAviz/Basic_Plot_Updated.py:
```diff
@@ -0,0 +1,130 @@
+"""This is a basic script for plotting an analysis file.
+
+    This script is just the first step in visualizing the data set. This file
```
It might be useful to let the user's know if there are any packages on which this code might be dependent on like  data_processing.py and probably a brief description of what inputs are being taken from the library.
A note on some new library packages  involved in the code like bokeh that might need to be installed before running the code.might be helpful too.

houghb replied:
Isn't it assumed that all pieces of a given package (in this case HDSAviz) are dependent on each other?  That will be especially clear by the set of import statements at the top of each file, so I don't think we need to bloat the code with extra comments about it - just my opinion.  There will be a list of all the project's dependencies in the final readme and documentation for the whole program, so I don't think we need to add it to every file...

bar commented:
@swapilpaliwal can you please stop using my handle to document your code? :) thanks!

cdf6gc committed:
Addressed more comments and added the toolbar

cdf6gc committed:
Added default values for inputs and ordered values

cdf6gc committed:
passed pep8

cdf6gc committed:
Checked last few easily implemented comments

cdf6gc commented:
I have addressed all of the comments listed above, with exception to implementing a stacked bar option or creating a hover tool. These will be included in the future tasks.
